src/main/java/org/opensearch/knn/indices/ModelCache.java

```diff
@@ -0,0 +1,134 @@
+import com.google.common.cache.CacheBuilder;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.cluster.service.ClusterService;
+
+import java.util.concurrent.ExecutionException;
+
+import static org.opensearch.knn.index.KNNSettings.MODEL_CACHE_SIZE_IN_BYTES_SETTING;
+
+
+public final class ModelCache {
+
+    private static Logger logger = LogManager.getLogger(ModelCache.class);
+
+    private static ModelCache instance;
+    private static ModelDao modelDao;
+    private static ClusterService clusterService;
+
+    private Cache<String, byte[]> cache;
+    private long cacheSizeInBytes;
+
+    /**
+     * Get instance of cache
+     *
+     * @return singleton instance of cache
+     */
+    public static synchronized ModelCache getInstance() {
+        if (instance == null) {
+            instance = new ModelCache();
+        }
+        return instance;
+    }
+
+    /**
+     * Initialize the cache
+     *
+     * @param modelDao modelDao used to read persistence layer for models
+     * @param clusterService used to update settings
+     */
+    public static void initialize(ModelDao modelDao, ClusterService clusterService) {
+        ModelCache.modelDao = modelDao;
+        ModelCache.clusterService = clusterService;
+    }
+
+    /**
+     * Evict all entries and rebuild the graph
+     */
+    public synchronized void rebuild() {
+        cache.invalidateAll();
+        initCache();
+    }
+
+    protected ModelCache() {
+        cacheSizeInBytes = MODEL_CACHE_SIZE_IN_BYTES_SETTING.get(clusterService.getSettings());
+        clusterService.getClusterSettings().addSettingsUpdateConsumer(MODEL_CACHE_SIZE_IN_BYTES_SETTING, it -> {
+            cacheSizeInBytes = it;
+            rebuild();
+        });
+        initCache();
+    }
+
+    private void initCache() {
+        CacheBuilder<String, byte[]> cacheBuilder = CacheBuilder.newBuilder()
+                .recordStats()
+                .concurrencyLevel(1)
+                .maximumWeight(cacheSizeInBytes)
+                .weigher((k, v) -> v.length);
+
+        cache = cacheBuilder.build();
+    }
+
+    /**
+     * Get the model from modelId
+     *
+     * @param modelId model identifier
+     * @return byte array representing model
+     */
+    public byte[] get(String modelId) {
+        try {
+            return cache.get(modelId, () -> modelDao.get(modelId));
+        } catch (ExecutionException ee) {
+            throw new IllegalStateException("Unable to retrieve model blob for \"" + modelId + "\": " + ee);
+        }
+    }
+
+    /**
+     * Get total weight of cache
+     *
+     * @return total weight
+     */
+    public long getTotalWeight() {
+        return cache.asMap().values().stream().map(bytes -> (long) bytes.length).reduce(0L, Long::sum);
+    }
+
+    /**
+     * Remove modelId from cache
+     *
+     * @param modelId to be removed
+     */
+    public void remove(String modelId) {
```

Comment:
shouldn't this be synchronized as well?

Reply:
I dont think so. All Guava cache implementations are expected to be thread safe: https://guava.dev/releases/17.0/api/docs/com/google/common/cache/Cache.html.

Reply:
This function eventually gets called: https://github.com/google/guava/blob/master/guava/src/com/google/common/cache/LocalCache.java#L3032-L3076